tests/UnitTests/Foundation/NSObject_NSKeyValueValidation.mm

```diff
@@ -0,0 +1,248 @@
+    EXPECT_OBJCEQ([NSData data], data);
+}
+
+TEST(NSObject, ValidateValueForKeyShouldReturnNOForInvalidSelectors) {
+    auto test = [ValidationObj test];
+    NSData* data = [NSData data];
+    NSError* error = nil;
+    EXPECT_EQ(NO, [test validateValue:&data forKey:@"valid1" error:&error]);
+    EXPECT_EQ(NO, [test validateValue:&data forKey:@"valid2" error:&error]);
+    EXPECT_EQ(NO, [test validateValue:&data forKey:@"valid3" error:&error]);
+    EXPECT_EQ(NO, [test validateValue:&data forKey:@"valid4" error:&error]);
+    EXPECT_EQ(NO, [test validateValue:&data forKey:@"valid5" error:&error]);
+    EXPECT_EQ(NO, [test validateValue:&data forKey:@"valid6" error:&error]);
+    EXPECT_EQ(NO, [test validateValue:&data forKey:@"valid7" error:&error]);
+    EXPECT_EQ(NO, [test validateValue:&data forKey:@"valid8" error:&error]);
+    EXPECT_EQ(NO, [test validateValue:&data forKey:@"valid9" error:&error]);
+    EXPECT_EQ(NO, [test validateValue:&data forKey:@"valid10" error:&error]);
+    EXPECT_EQ(NO, [test validateValue:&data forKey:@"valid11" error:&error]);
+    EXPECT_EQ(NO, [test validateValue:&data forKey:@"valid12" error:&error]);
+    EXPECT_EQ(NO, [test validateValue:&data forKey:@"valid13" error:&error]);
+    EXPECT_EQ(NO, [test validateValue:&data forKey:@"valid14" error:&error]);
+    EXPECT_EQ(NO, [test validateValue:&data forKey:@"valid15" error:&error]);
+    EXPECT_EQ(NO, [test validateValue:&data forKey:@"valid16" error:&error]);
+    EXPECT_EQ(NO, [test validateValue:&data forKey:@"valid17" error:&error]);
+
+    EXPECT_EQ(YES, [test validateValue:&data forKey:@"invalid1" error:&error]);
+    EXPECT_EQ(YES, [test validateValue:&data forKey:@"invalid2" error:&error]);
+    EXPECT_EQ(YES, [test validateValue:&data forKey:@"invalid3" error:&error]);
+
+    EXPECT_EQ(YES, [test validateValue:&data forKey:@"nonexistent" error:&error]);
+}
+
+@interface ValidationHolder : NSObject
++ (instancetype)holder;
+@property (readwrite, retain) ValidationObj* obj;
+@end
+
+@implementation ValidationHolder
+
++ (instancetype)holder {
+    return [[[self alloc] init] autorelease];
+}
+
+- (instancetype)init {
+    if (self = [super init]) {
+        _obj = [[ValidationObj alloc] init];
+    }
+
+    return self;
+}
+
+- (void)dealloc {
+    [_obj release];
+    [super dealloc];
+}
+@end
+
+TEST(NSObject, ValidateValueForKeyPath) {
+    ValidationHolder* test = [ValidationHolder holder];
+    NSData* data = nil;
+    NSError* error = nil;
+
+    EXPECT_EQ(NO, [test validateValue:nil forKeyPath:@"obj.obj" error:&error]);
+    EXPECT_EQ(YES, [test validateValue:&data forKeyPath:@"obj.obj" error:&error]);
+    EXPECT_OBJCEQ([NSData data], data);
+}
+
+TEST(NSObject, ValidateValueForKeyPathShouldProperlyFollowKeyPath) {
+    ValidationHolder* test = [ValidationHolder holder];
+    NSData* data = [NSData data];
+    NSError* error = nil;
+    EXPECT_EQ(NO, [test validateValue:&data forKeyPath:@"obj.valid1" error:&error]);
```

Comment:
I wasn't able to get it called by value setters, and documentation seems to suggest it's something that consumers use manually, so I don't believe so